# Data Cleaning and Preprocessing

In [1]:
!pip install TextBlob

In [2]:
import pandas as pd
import numpy as np
import isodate
import ast
from isodate import parse_duration
from textblob import TextBlob

In [3]:
video_df = pd.read_csv("dataFolder/raw/redPillAnalytics.csv")

In [4]:
df1 = pd.read_csv("dataFolder/raw/commentsBatchOne_df.csv")
df2 = pd.read_csv("dataFolder/raw/commentsBatchTwo_df.csv")
comments_df = pd.concat([df1, df2], ignore_index=True)

In [5]:
df_filtered = video_df[['video_id', 'channelTitle']]

In [6]:
comments_df = df_filtered.merge(comments_df, on='video_id', how='inner').copy()
comments_df

video_id                     channelTitle  \
0       F5eSaabAAmk                    Benjamin Seda   
1       F5eSaabAAmk                    Benjamin Seda   
2       F5eSaabAAmk                    Benjamin Seda   
3       F5eSaabAAmk                    Benjamin Seda   
4       F5eSaabAAmk                    Benjamin Seda   
...             ...                              ...   
668718  Jyjqw_HwXVg  The Corbett Report (Unofficial)   
668719  Jyjqw_HwXVg  The Corbett Report (Unofficial)   
668720  Jyjqw_HwXVg  The Corbett Report (Unofficial)   
668721  Jyjqw_HwXVg  The Corbett Report (Unofficial)   
668722  Jyjqw_HwXVg  The Corbett Report (Unofficial)   

                                                  comment  \
0                                          big boobs lmao   
1       This will work for a specific type of woman, o...   
2       Can you do a video on what to do  if you  enco...   
3                                      God of the Dates 🤍   
4       About cold approaches, it's just not true. I d...   
...                                                   ...   
668718  Make everyone you know, or are even slightly r...   
668719                   The Green Scheme is Green Death!   
668720            a real timeless classic, perfect choice   
668721                                This is great stuff   
668722    Did NOT expect this notification hahaha amazing   

                published_at  
0       2025-03-10T00:36:43Z  
1       2025-03-09T23:01:56Z  
2       2025-03-09T07:13:54Z  
3       2025-03-08T14:57:50Z  
4       2025-03-07T19:06:08Z  
...                      ...  
668718  2022-05-25T12:19:52Z  
668719  2022-05-25T12:15:15Z  
668720  2022-05-23T18:30:31Z  
668721  2022-05-23T16:13:58Z  
668722  2022-05-23T15:46:22Z  

[668723 rows x 4 columns]

In [7]:
comments_df.isnull().any()

video_id        False
channelTitle    False
comment          True
published_at    False
dtype: bool

In [8]:
comments_df.dtypes

video_id        object
channelTitle    object
comment         object
published_at    object
dtype: object

In [9]:
comments_df.isnull().sum()
comments_df.dropna()
comments_df.isin([0, np.nan]).sum()

video_id         0
channelTitle     0
comment         51
published_at     0
dtype: int64

In [10]:
comments_df = comments_df.dropna(subset=['comment'])
print(comments_df.isnull().sum())  # Verify no nulls in 'comment'

video_id        0
channelTitle    0
comment         0
published_at    0
dtype: int64


In [11]:
video_df.isnull().any()

video_id          False
channelTitle      False
title             False
description        True
tags               True
publishedAt       False
viewCount         False
likeCount          True
favouriteCount     True
commentCount       True
duration          False
definition        False
caption           False
dtype: bool

In [12]:
video_df.dtypes

video_id           object
channelTitle       object
title              object
description        object
tags               object
publishedAt        object
viewCount           int64
likeCount         float64
favouriteCount    float64
commentCount      float64
duration           object
definition         object
caption              bool
dtype: object

In [13]:
video_df.isnull().sum()
video_df.dropna()
video_df.isin([0, np.nan]).sum()

video_id              0
channelTitle          0
title                 0
description         394
tags               6314
publishedAt           0
viewCount             3
likeCount            23
favouriteCount    13675
commentCount        150
duration              0
definition            0
caption           13246
dtype: int64

In [14]:
video_df.isnull().sum()

video_id              0
channelTitle          0
title                 0
description         394
tags               6314
publishedAt           0
viewCount             0
likeCount            16
favouriteCount    13675
commentCount          9
duration              0
definition            0
caption               0
dtype: int64

In [15]:
video_df['description'] = video_df['description'].fillna("")

# Convert string representations of tags to lists and fill missing values with empty lists
def safe_literal_eval(x):
    try:
        return ast.literal_eval(x) if isinstance(x, str) else []
    except (ValueError, SyntaxError):
        return []

video_df['tags'] = video_df['tags'].apply(safe_literal_eval)

# Fill missing likeCount, favouriteCount, and commentCount with 0
video_df['likeCount'] = video_df['likeCount'].fillna(0)
video_df['favouriteCount'] = video_df['favouriteCount'].fillna(0)
video_df['commentCount'] = video_df['commentCount'].fillna(0)

# Verify no null values remain
print(video_df.isnull().sum())

video_id          0
channelTitle      0
title             0
description       0
tags              0
publishedAt       0
viewCount         0
likeCount         0
favouriteCount    0
commentCount      0
duration          0
definition        0
caption           0
dtype: int64


In [16]:
print(video_df['tags'].head())

0    [how to flirt with a girl, dates, how to get a...
1    [how to flirt with a girl, dates, how to get a...
2    [how to flirt with a girl, dates, how to get a...
3    [how to flirt with a girl, dates, how to get a...
4    [how to flirt with a girl, dates, how to get a...
Name: tags, dtype: object


In [17]:
numeric_cols = ['viewCount', 'likeCount', 'favouriteCount', 'commentCount']
video_df[numeric_cols] = video_df[numeric_cols].apply(pd.to_numeric, errors = 'coerce', axis = 1)

In [18]:
# Convert publishedAt to datetime
video_df['publishedAt'] = pd.to_datetime(video_df['publishedAt']).dt.tz_localize(None)

# Extract day of the week
video_df['publishDayName'] = video_df['publishedAt'].dt.strftime("%A")

# Convert publishedAt to Unix timestamp
video_df['publishedAt_timestamp'] = video_df['publishedAt'].astype('int64') / 10**9

# Count tags
video_df['tagCount'] = video_df['tags'].apply(lambda x: len(x) if isinstance(x, list) else [])

# Convert duration to seconds
video_df['durationSecs'] = video_df['duration'].apply(lambda x: isodate.parse_duration(x).total_seconds())

video_df['durationSecs'] = video_df['duration'].apply(
    lambda x: isodate.parse_duration(x).total_seconds() if pd.notnull(x) else 0
)

# Calculate title length
video_df['titleLength'] = video_df['title'].apply(lambda x: len(x))

In [19]:
video_df

video_id                     channelTitle  \
0      F5eSaabAAmk                    Benjamin Seda   
1      xJ6b8CV-pQ0                    Benjamin Seda   
2      kPhrei5S88U                    Benjamin Seda   
3      4ZnwTwLcAeM                    Benjamin Seda   
4      VW9-SBs6yIg                    Benjamin Seda   
...            ...                              ...   
13670  _X4P6L622-8  The Corbett Report (Unofficial)   
13671  tkmZ4c2AOVY  The Corbett Report (Unofficial)   
13672  mr7itEUIVew  The Corbett Report (Unofficial)   
13673  ochRNyIDTE8  The Corbett Report (Unofficial)   
13674  Jyjqw_HwXVg  The Corbett Report (Unofficial)   

                                                   title  \
0      How to ACTUALLY Get a Girlfriend in 2025 (Full...   
1                         How to Find A 10/10 Girlfriend   
2      The Mistake 99% of Men Make That Keep Them Single   
3                        How to Always Get That 2nd Date   
4        The Donald Trump Method for Tinder (STEAL THIS)   
...                                                  ...   
13670  How Do I Defend Voluntarism? - Questions For C...   
13671                                     The 5G Dragnet   
13672  False Flags: The Secret History of Al Qaeda — ...   
13673  Episode 409 - False Flags: The Secret History ...   
13674                    HOW BIG OIL CONQUERED THE WORLD   

                                             description  \
0      👉🏼 Get 1-3+ dates per week in 30 days (coachin...   
1      👫 My 3 step formula to approach & attract wome...   
2      👫 My 3 step formula to approach & attract wome...   
3      👫 My 3 step formula to approach & attract wome...   
4      👫 My 3 step formula to approach & attract wome...   
...                                                  ...   
13670  I have no affiliation with James Corbett or Th...   
13671  I have no affiliation with James Corbett or Th...   
13672  I have no affiliation with James Corbett or Th...   
13673  I have no affiliation with James Corbett or Th...   
13674  I have no affiliation with James Corbett or Th...   

                                                    tags         publishedAt  \
0      [how to flirt with a girl, dates, how to get a... 2025-03-06 15:27:49   
1      [how to flirt with a girl, dates, how to get a... 2025-03-03 15:01:24   
2      [how to flirt with a girl, dates, how to get a... 2025-03-01 14:45:07   
3      [how to flirt with a girl, dates, how to get a... 2025-02-27 14:15:00   
4      [how to flirt with a girl, dates, how to get a... 2025-02-26 13:45:03   
...                                                  ...                 ...   
13670                                                 [] 2022-05-25 13:22:06   
13671                                                 [] 2022-05-24 17:28:15   
13672                                                 [] 2022-05-24 17:01:10   
13673                                                 [] 2022-05-24 17:01:06   
13674                                                 [] 2022-05-23 15:45:56   

       viewCount  likeCount  favouriteCount  commentCount    duration  \
0         5034.0      254.0             0.0          27.0     PT15M4S   
1         3346.0      330.0             0.0          22.0       PT59S   
2         2690.0      222.0             0.0          19.0       PT36S   
3         4060.0      413.0             0.0           9.0       PT46S   
4         6818.0      316.0             0.0          30.0       PT32S   
...          ...        ...             ...           ...         ...   
13670      393.0       15.0             0.0           2.0    PT28M13S   
13671     1226.0       61.0             0.0           2.0    PT25M44S   
13672     5531.0      179.0             0.0          16.0  PT1H16M19S   
13673     6132.0      212.0             0.0          16.0  PT1H59M39S   
13674     8782.0      340.0             0.0          30.0  PT1H11M27S   

      definition  caption publishDayName  publishedAt_timestamp  tagC

In [20]:
channel_stats = video_df.groupby('channelTitle').agg({
    'viewCount': 'sum',
    'likeCount': 'sum',
    'commentCount': 'sum'
}).reset_index()

In [21]:
channel_stats['engagementRate'] = (channel_stats['likeCount'] + channel_stats['commentCount']) / channel_stats['viewCount']

In [22]:
channel_stats

channelTitle    viewCount   likeCount  commentCount  \
0                       Alhpamales        239.0        10.0           0.0   
1                    Benjamin Seda  202773449.0   6456129.0      356756.0   
2                  Better Bachelor  231797826.0  12936057.0     2997777.0   
3                Coach Corey Wayne  251196988.0   4876107.0      396843.0   
4                      FreshandFit  221150653.0   9730118.0      819214.0   
5                Jordan B Peterson  950219566.0  31148819.0     3469047.0   
6     Paul Joseph Watson | Перевод       2178.0        44.0           0.0   
7  The Corbett Report (Unofficial)    1124728.0     84132.0       12672.0   
8                 The Distributist    6873745.0    248340.0       92768.0   

   engagementRate  
0        0.041841  
1        0.033599  
2        0.068740  
3        0.020991  
4        0.047702  
5        0.036431  
6        0.020202  
7        0.086069  
8        0.049625

## Breakdown of Channels

| Channel                        | View Count   | Engagement Rate | Relevance to Red Pill                              |
|--------------------------------|--------------|-----------------|---------------------------------------------------|
| Benjamin Seda                  | 202,773,449  | 3.36%           | Likely relevant (focus on dating, masculinity, and self-improvement) |
| Better Bachelor                | 231,797,826  | 6.87%           | Highly relevant (focus on men’s rights, dating, and anti-feminism) |
| Coach Corey Wayne              | 251,196,988  | 2.10%           | Relevant (focus on dating advice and relationships) |
| FreshandFit                    | 221,150,653  | 4.77%           | Highly relevant (focus on dating, gender dynamics, and masculinity) |
| Jordan B Peterson              | 950,219,566  | 3.64%           | Partially relevant (focus on psychology, self-improvement, and gender roles) |
| The Corbett Report (Unofficial)| 1,124,728    | 8.61%           | Less relevant (focus on conspiracy theories, not directly Red Pill) |
| The Distributist               | 6,873,745    | 4.96%           | Less relevant (focus on traditionalism and economics, not directly Red Pill) |
| Alhpamales                     | 239          | 4.18%           | Likely irrelevant (very low view count, unclear relevance) |
| Paul Joseph Watson | Перевод     | 2,178          | 2.02%           | Likely irrelevant (low view count, unclear relevance) |

## Recommendations

### Channels to Keep (Highly Relevant to Red Pill):
- **Better Bachelor**
- **FreshandFit**
- **Benjamin Seda**
- **Coach Corey Wayne**
- **Jordan B Peterson** (if you consider his content relevant to Red Pill themes)

### Channels to Exclude (Less Relevant or Low Engagement):
- **The Corbett Report (Unofficial)** (focuses on conspiracy theories, not Red Pill)
- **The Distributist** (focuses on traditionalism/economics, not Red Pill)
- **Alhpamales** and **Paul Joseph Watson | Перевод** (very low engagement, unclear relevance)

In [23]:
excluded_channels = [
    "The Corbett Report (Unofficial)", 
    "The Distributist", 
    "Alhpamales", 
    "Paul Joseph Watson | Перевод"
]

video_df = video_df[~video_df["channelTitle"].isin(excluded_channels)].copy()

In [19]:
comments_df = comments_df[~comments_df["channelTitle"].isin(excluded_channels)].copy()

In [24]:
video_df.groupby('channelTitle')['durationSecs'].mean()

channelTitle
Benjamin Seda         195.985205
Better Bachelor      1379.782349
Coach Corey Wayne     659.063435
FreshandFit          3706.578228
Jordan B Peterson    2940.362724
Name: durationSecs, dtype: float64

In [25]:
# Calculate view_per_like (handle division by zero)
video_df['view_per_like'] = np.where(
    video_df['likeCount'] != 0,  # Condition: likeCount is not zero
    video_df['viewCount'] / video_df['likeCount'],  # True: Perform division
    0  # False: Set to 0 if likeCount is zero
)

# Calculate comment_duration_interaction
video_df['comment_duration_interaction'] = video_df['commentCount'] * video_df['durationSecs']

# Calculate popularity_score
video_df['popularity_score'] = (
    video_df['viewCount'] + 
    video_df['likeCount'] * 10 + 
    video_df['commentCount'] * 20
)

# Calculate commentRatio (handle division by zero)
video_df['commentRatio'] = np.where(
    video_df['viewCount'] != 0,  # Condition: viewCount is not zero
    video_df['commentCount'] / video_df['viewCount'],  # True: Perform division
    0  # False: Set to 0 if viewCount is zero
)

# Calculate likeRatio (handle division by zero)
video_df['likeRatio'] = np.where(
    video_df['viewCount'] != 0,  # Condition: viewCount is not zero
    video_df['likeCount'] / video_df['viewCount'],  # True: Perform division
    0  # False: Set to 0 if viewCount is zero
)

# Fill any remaining NaN values with 0 (if needed)
video_df = video_df.fillna(0)

In [26]:
video_df['title_sentiment'] = video_df['title'].apply(lambda x: TextBlob(x).sentiment.polarity)

In [27]:
video_df['commentCount'].mean()

623.0344854308742

In [28]:
video_df[['viewCount', 'commentCount']].corr()

viewCount  commentCount
viewCount      1.000000      0.617668
commentCount   0.617668      1.000000

In [29]:
video_df

video_id   channelTitle  \
0      F5eSaabAAmk  Benjamin Seda   
1      xJ6b8CV-pQ0  Benjamin Seda   
2      kPhrei5S88U  Benjamin Seda   
3      4ZnwTwLcAeM  Benjamin Seda   
4      VW9-SBs6yIg  Benjamin Seda   
...            ...            ...   
12917  KO7Z0HdxIek    FreshandFit   
12918  3LXo6A-JnV4    FreshandFit   
12919  e9Gdl-szTg4    FreshandFit   
12920  RHlPDYsuBYs    FreshandFit   
12921  FirJFTPpE8M    FreshandFit   

                                                   title  \
0      How to ACTUALLY Get a Girlfriend in 2025 (Full...   
1                         How to Find A 10/10 Girlfriend   
2      The Mistake 99% of Men Make That Keep Them Single   
3                        How to Always Get That 2nd Date   
4        The Donald Trump Method for Tinder (STEAL THIS)   
...                                                  ...   
12917            The Most OPTIMAL Rep Range For STRENGTH   
12918  What's better for fat loss? Low carb or high c...   
12919  Are Fitness/Calorie Tracking Apps Accurate? Th...   
12920      IS FASTING SUPERIOR? What the science says...   
12921                 Welcome to the TRUTH in Fitness...   

                                             description  \
0      👉🏼 Get 1-3+ dates per week in 30 days (coachin...   
1      👫 My 3 step formula to approach & attract wome...   
2      👫 My 3 step formula to approach & attract wome...   
3      👫 My 3 step formula to approach & attract wome...   
4      👫 My 3 step formula to approach & attract wome...   
...                                                  ...   
12917  Many people struggle to determine the best rep...   
12918  Today we answer the age old question. Are high...   
12919  Are these popular apps bringing you closer to ...   
12920  Does fasting build more muscle or help burn mo...   
12921  Don't forget to LIKE, SUBSCRIBE, and COMMENT! ...   

                                                    tags         publishedAt  \
0      [how to flirt with a girl, dates, how to get a... 2025-03-06 15:27:49   
1      [how to flirt with a girl, dates, how to get a... 2025-03-03 15:01:24   
2      [how to flirt with a girl, dates, how to get a... 2025-03-01 14:45:07   
3      [how to flirt with a girl, dates, how to get a... 2025-02-27 14:15:00   
4      [how to flirt with a girl, dates, how to get a... 2025-02-26 13:45:03   
...                                                  ...                 ...   
12917  [repetition range, how to lift weights, how to... 2020-06-25 00:36:00   
12918  [fatloss, lowcarb, weight loss, keto, evidence... 2020-06-20 14:45:11   
12919                 [redpill fitness hypergamy #gainz] 2020-06-13 15:00:11   
12920                      [fasting, fitness, aesthetic] 2020-05-30 15:00:27   
12921                                 [Fitness, Redpill] 2020-05-08 15:00:10   

       viewCount  likeCount  favouriteCount  commentCount  ...  \
0         5034.0      254.0             0.0          27.0  ...   
1         3346.0      330.0             0.0          22.0  ...   
2         2690.0      222.0             0.0          19.0  ...   
3         4060.0      413.0             0.0           9.0  ...   
4         6818.0      316.0             0.0          30.0  ...   
...          ...        ...             ...           ...  ...   
12917     9633.0      601.0             0.0          21.0  ...   
12918     9209.0      490.0             0.0          26.0  ...   
12919     4787.0      242.0             0.0          10.0  ...   
12920    23952.0     1353.0             0.0          64.0  ...   
12921    18681.0      542.0             0.0          49.0  ...   

      publishedAt_timestamp tagCount  durationSecs titleLength  view_per_like  \
0              1.741275e+09       21         904.0          53      19.818898   
1              1.741014e+09       13          59.0          30      10.139394   
2              1.740840e+09       13          36.0          49      12.117117   
3              1.740666e+09       13         

In [30]:
top_10_views = video_df.nlargest(10, 'viewCount')[['title', 'viewCount', 'channelTitle']]
print(top_10_views)

                                                   title   viewCount  \
9910           JBP X @MattRifeComedy.  Today at 5pm EST.  18502980.0   
10451  Lecture: Biblical Series I: Introduction to th...  13706650.0   
1463   THIS is How A Girl Wants You to TEXT HER | How...  11807649.0   
1473   7 Ways To INSTANTLY Look MORE ATTRACTIVE | How...   8246244.0   
9984                             COVID-19 Cause of Death   7952713.0   
10053  Talking to Muslims About Christ | Mohammed Hij...   7756906.0   
9973                            What Are Women Good For?   7656196.0   
10072         Africa is Not Poor Because of Colonization   7541337.0   
10389  Documentary: A Glitch in the Matrix (David Ful...   6605578.0   
90             If a Girl is Looking at You, Approach Her   6594217.0   

            channelTitle  
9910   Jordan B Peterson  
10451  Jordan B Peterson  
1463       Benjamin Seda  
1473       Benjamin Seda  
9984   Jordan B Peterson  
10053  Jordan B Peterson  
9973   Jordan B Pe

In [31]:
top_10_likes = video_df.nlargest(10, 'likeCount')[['title', 'likeCount', 'channelTitle']]
print(top_10_likes)

                                                   title  likeCount  \
9910           JBP X @MattRifeComedy.  Today at 5pm EST.   611913.0   
10290                                        Return Home   543720.0   
9955   The Fight Against Worldwide Child Slavery & th...   318009.0   
10110                               Article: Twitter Ban   315518.0   
9973                            What Are Women Good For?   305820.0   
9984                             COVID-19 Cause of Death   299559.0   
10087    Language Is Used as a Group Protection Strategy   232440.0   
114                 If A Girl is Looking at You, Do This   231402.0   
10451  Lecture: Biblical Series I: Introduction to th...   227120.0   
10072         Africa is Not Poor Because of Colonization   220386.0   

            channelTitle  
9910   Jordan B Peterson  
10290  Jordan B Peterson  
9955   Jordan B Peterson  
10110  Jordan B Peterson  
9973   Jordan B Peterson  
9984   Jordan B Peterson  
10087  Jordan B Peterson  
11

In [32]:
video_df.to_csv("dataFolder/processed/cleanedDataFrame.csv", index=False)

In [21]:
comments_df.to_csv("dataFolder/processed/cleanedComments.csv", index=False)